runtime/runtime/src/congestion_control.rs

```diff
@@ -223,37 +227,80 @@ impl ReceiptSinkV2 {
         &mut self,
         state_update: &mut TrieUpdate,
         apply_state: &ApplyState,
+        epoch_info_provider: &dyn EpochInfoProvider,
     ) -> Result<(), RuntimeError> {
-        // store shards in vec to avoid borrowing self.outgoing_limit
-        let shards: Vec<_> = self.outgoing_limit.keys().copied().collect();
-        for shard_id in shards {
-            self.forward_from_buffer_to_shard(shard_id, state_update, apply_state)?;
+        tracing::debug!(target: "runtime", "forwarding receipts from outgoing buffers");
+
+        // TODO(wacban) - check if the order is the same as previously, if not
+        // then this is a protocol upgrade and needs to be handled as such.
+        // DO NOT COMMIT until this is done.
+        let shard_layout = epoch_info_provider.shard_layout(&apply_state.epoch_id)?;
+        let shard_ids = shard_layout.shard_ids().collect::<Vec<_>>();
+
+        // TODO(resharding) HashSet is non-deterministic, this will break if there
+        // are multiple parents. This may happen if there are multiple splits
+        // within the same resharding.
+        let mut parent_shard_ids = HashSet::new();
+        for &shard_id in &shard_ids {
+            let parent_shard_id = shard_layout.try_get_parent_shard_id(shard_id).unwrap();
+            let Some(parent_shard_id) = parent_shard_id else {
+                continue;
+            };
+            if parent_shard_id == shard_id {
+                continue;
+            }
+            parent_shard_ids.insert(parent_shard_id);
+        }
+
+        // First forward any receipts that may still be in the outgoing buffers
+        // of the parent shards.
+        for &shard_id in &parent_shard_ids {
+            self.forward_from_buffer_to_shard(shard_id, state_update, apply_state, &shard_layout)?;
+        }
+
+        // Then forward receipts from the outgoing buffers of the shard in the
+        // current shard layout.
+        //
+        // TODO(wacban) This doesn't seem fair. Or is it?
```

Comment:
Do you think forwarding receipts up `limit / number of shards + 1(to count parent)`  for each shard would make it more fair? 
Although I can't really picture how much we are interested in fairness here..

Reply:
I was wondering if the processing order matter. Here we always go through the receipts in the order defined by the shard layout. I'm thinking if this gives the shards at lower indices any advantage. It may not be the case if the limits are independent but I just want to have another look at it later to get some understanding.

Reply:
I think it's fair since the limits for each shard are independent. Removing this todo.